### ISSUE TITLE
Error -3 while decompressing data: invalid stored block lengths 

### ISSUE THREAD

**Issue by ankasani** [python/cpython#128258]
# Bug report

### Bug description:

When using below code is working fine with 3.11 version and not working with 3.12, 3.13.

```
import zipfile
import asyncio

async def process_file(text_file_name: str, zip_file: zipfile.ZipFile):
    try:
        with zip_file.open(text_file_name, mode='r') as text_file:
            try:
                content = await asyncio.to_thread(text_file.read)
                lines = content.decode('utf-8').splitlines()
            except UnicodeDecodeError as e:
                print(f"Error decoding file {text_file_name}: {e}")
                return None
            except Exception as e:
                print(f"Error reading file {text_file_name}: {e}")
                return None
            if not lines:
                return None
            # Process lines here
            return lines
    except Exception as e:
        print(f"Error opening file {text_file_name}: {e}")
        return None

async def main():
    temp_file_path = 'Tests.zip'
    zip_file = zipfile.ZipFile(temp_file_path, 'r')
    tasks = [process_file(text_file_name, zip_file) for text_file_name in zip_file.namelist()]
    await asyncio.gather(*tasks)

asyncio.run(main())
```
Also, open the ZIP file inside the process_file function, it was leading to higher memory usage. This is because each task would open a new instance of the ZIP file, potentially loading multiple instances into memory simultaneously, especially if the ZIP contains many files. I’m looking for a solution that minimizes memory usage while still allowing for fast, simultaneous execution. so any one has any suggestions on it ?

### CPython versions tested on:

3.11, 3.12, 3.13

### Operating systems tested on:

Linux, Windows

**Comment by picnixz** [comment]
Can you bisect the bad commit? either something in asyncio or in zipfile changed (though I'm not really sure we support concurrent read accesses)

**Comment by picnixz** [comment]
To bisect, you should compile Python from sources and find which commit introduced this change. However, if you're not familiar with this, I or someone else will do it.

**Comment by picnixz** [comment]
We can fix the issue for 3.12, 3.13 and 3.14 indeed since those are still bug-fixable. But I'm not sure we want to do it (are there more use cases when this would help or not?) Since I don't have an answer to that question, I'll ask those that approved the PR (@FFY00 and @corona10) and some other people that are involved in asyncio, zipfile and IO perfs (@kumaraditya303 @graingert @jaraco @serhiy-storchaka)

**Comment by graingert** [comment]
It doesn't look like ZipFile is thread safe at all, I think it's just luck that this was working on 3.11

**Comment by danifus** [comment]
@ankasani looks like you were right that it was related to https://github.com/python/cpython/issues/127847! Well spotted :p 

https://github.com/python/cpython/issues/127847 is a bug in `_SharedFile.seek()` when using `whence=os.SEEK_CUR` (also known as `whence=1`) which is what the change you bisected to introduced. That bug fix has been merged 2 days ago so you should check if this is resolved on main (or the 3.13 or 3.12 branches which got the backport too).

Great work tracking down the cause :)

**Comment by picnixz** [comment]
Since it's working now, I'm going to close it as a `duplicate` of https://github.com/python/cpython/issues/127847. However, I thank you for your patience and the fact that you bisected the commit yourself. Concerning the thread-safety of `ZipFile` I think this should be a separated issue.
